### PR TITLE
Improve far seek performance of chunkSampleStream

### PR DESCRIPTION
**Pull request by WeiChungChang**
Rework to make the patch accordingly by @ojw28 kindly suggestions #as #2247.

Issue report: #2253

### CODE REVIEW

File: library/src/main/java/com/google/android/exoplayer2/source/chunk/ChunkSampleStream.java [hunk 1]
```diff
@@ -115,14 +115,29 @@ public long getBufferedPositionUs() {
...
    * Seeks to the specified position in microseconds.
    *
    * @param positionUs The seek position in microseconds.
    */
   public void seekToUs(long positionUs) {
     lastSeekPositionUs = positionUs;
     // If we're not pending a reset, see if we can seek within the sample queue.
-    boolean seekInsideBuffer = !isPendingReset() && sampleQueue.skipToKeyframeBefore(positionUs);
+    boolean seekInsideBuffer = !isPendingReset() && 
+        sampleQueue.skipToKeyframeBefore(positionUs, isWithinLastChunk(positionUs));
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer2/source/chunk/ChunkSampleStream.java]: I'm not sure you need to be as sophisticated as checking whether the position is within the last chunk. Isn't it sufficient just to pass `positionUs < getNextLoadPositionUs()` as the second argument? I don't think passing true does any harm in cases where `positionUs` is *earlier* than the last chunk.

This allows you to remove your `isWithinLastChunk` and `getCurrentLoadPosition` methods ;).

File: library/src/main/java/com/google/android/exoplayer2/extractor/DefaultTrackOutput.java [hunk 2]
```diff
@@ -229,10 +229,26 @@ public long getLargestQueuedTimestampUs() {
...
+
+  /**
+   * Attempts to skip to the keyframe before the specified time, if it's present in the buffer.
+   *
+   * @param timeUs The seek time.
    * @return Whether the skip was successful.
    */
   public boolean skipToKeyframeBefore(long timeUs) {
-    long nextOffset = infoQueue.skipToKeyframeBefore(timeUs);
+    long nextOffset = infoQueue.skipToKeyframeBefore(timeUs, false);
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer2/extractor/DefaultTrackOutput.java]: You can replace the whole body of this method with just:
`return skipToKeyframeBefore(timeUs, false)`

**Review by ojw28 — COMMENTED** [review]
Looking good; thanks! I think you can make a few further simplifications, but otherwise think we're good to get this merged. Please take a look.

**Comment by WeiChungChang** [comment]
Dear @ojw28:
Thanks a lot for your suggestion to make it clearer and simpler.
I have made the new patch as 'refine_to_make_flow_simpler':
https://github.com/google/ExoPlayer/pull/2318/commits/2214a4c152fab862c928875f8265e77676d3652b
Thank you!

**Pull request merged**